uber/site_sections/preregistration.py

```diff
@@ -562,10 +562,25 @@ def invalid_badge(self, session, id, message=''):
     def not_found(self, id, message):
         return {'id': id, 'message': message}
 
-    def invalidate(self, session, id):
+    def abandon_badge(self, session, id):
         attendee = session.attendee(id)
-        attendee.badge_status = c.INVALID_STATUS
-        raise HTTPRedirect('invalid_badge?id={}&message={}', attendee.id, 'Sorry you can\'t make it! We hope to see you next year!')
+        success_message = "Sorry you can't make it! We hope to see you next year!"
+        failure_message = "You cannot abandon your badge because you are the leader of a group."
+
+        # if attendee is part of a group, we must delete attendee and remove from the group
+        if attendee.group:
```

Comment:
Minor suggestion to clean this block up a little--

Since `raise`ing an HTTPRedirect prevents the rest of the function from running, it will be easy and may look nicer to rearrange the code so that we first check if they are a group leader and return a 'failure' message, and then we handle the 'success' cases. So that would look like:
```
if attendee.is_group_leader:
  # attendee cannot invalidate their badge, as they are the leader of a group
  raise HTTPRedirect('confirm?id={}&message={}', id, failure_message)

### Now check if they're in a group vs not to determine if we delete the badge or invalidate it
```

Both styles work and make sense so if you want to keep this as-is, that's fine!

By the way, you'll note the use of `attendee.is_group_leader` here. This is a convenience property we have (https://github.com/magfest/ubersystem/blob/master/uber/models.py#L1792) that helps keep us from having to do checks like `if attendee.group and attendee.group.leader == attendee` all the time. I definitely recommend using that. We have a lot of these kinds of properties, they're all in models.py -- basically, if it's some kind of attendee or group status we're liable to check a lot, it probably has a custom property associated with it.

Reply:
I appreciate the advice, your suggestion to handle the error case first does make more sense. I will make this change.

I also didn't realize there was such a property, but of course there would be. I should take the time to review models.py. Thanks!